dingo/preprocess.py

```diff
@@ -0,0 +1,245 @@
+        self._model.reactions.get_by_id(self._objective).lower_bound = fba_solution.objective_value
+
+        fva = cobra.flux_analysis.flux_variability_analysis(self._model, fraction_of_optimum=0.95)
+        mle = fva.loc[ (abs(fva['minimum']) < tol ) & (abs(fva['maximum']) < tol)]
+        mle = mle.index.tolist()
+        
+        self._model.reactions.get_by_id(self._objective).lower_bound = wt_lower_bound
+        
+        self._mle_reactions = mle
+        return self._mle_reactions
+    
+    
+    def _remove_model_reactions(self):
+        """
+        A function used to set the lower and upper bounds of certain reactions to 0
+        (it turns off reactions)
+        """
+                                
+        for reaction in self._removed_reactions:
+            self._model.reactions.get_by_id(reaction).lower_bound = 0
+            self._model.reactions.get_by_id(reaction).upper_bound = 0
+            
+        return self._model
+   
+            
+    def reduce(self, extend):
+        """
+        A function that calls the "remove_model_reactions" function
+        and removes blocked, zero-flux and metabolically less efficient 
+        reactions from the model.
+        
+        Then it finds the remaining reactions in the model after 
+        exclusion of the essential reactions.
+        
+        When the "extend" parameter is set to 1, the function  performes
+        an additional check to remove further reactions. These reactions 
+        are the ones that if knocked-down, they do not affect the value 
+        of the objective function. These reactions are removed simultaneously
+        from the model. If this simultaneous removal produces an infesible
+        solution (or a solution of 0) to the objective function, 
+        these reactions are restored with their initial bounds.
+        
+        A dingo-type tuple is then created from the cobra model 
+        using the "cobra_dingo_tuple" function.
+        
+        The outputs are
+        (a) A list of the removed reactions ids
+        (b) A reduced dingo model
+        """        
+        
+        # create a list from the combined blocked, zero-flux, mle reactions
+        blocked_mle_zero = self._blocked_reactions + self._mle_reactions + self._zero_flux_reactions
+        list_removed_reactions = list(set(blocked_mle_zero))        
+        self._removed_reactions = list_removed_reactions
+   
+        # remove these reactions from the model
+        self._remove_model_reactions()
+                     
+        remained_reactions = list((Counter(self._initial_reactions)-Counter(self._removed_reactions)).elements())
+        remained_reactions = list((Counter(remained_reactions)-Counter(self._essential_reactions)).elements())
+        
+        tol = 1e-6
+        
+        if extend != 0 and extend != 1:
+            raise Exception("Wrong Input to extend parameter")
+  
+        # find additional reactions with a possibility of removal
```

Comment:
It seems to me you need to move the `extend` check here.
I do not see a reason to have any of the loop for the cases when `extend` is `False`.